packages/core/application/core-application-browser-internal/src/application_service.tsx

```diff
@@ -406,9 +406,14 @@ export class ApplicationService {
 
   private async shouldNavigate(overlays: OverlayStart, nextAppId: string): Promise<boolean> {
     const currentAppId = this.currentAppId$.value;
-    if (currentAppId === undefined) {
+    if (
+      currentAppId === undefined ||
+      (currentAppId === 'dashboards' &&
+        (nextAppId === 'discover' || nextAppId === 'visualize' || nextAppId === 'lens'))
+    ) {
```

Comment:
I'm not sure this is the best solution - having all of these hard-coded values seems... risky 🙈 Is there a reason you didn't implement this change via the app leave handler in `src/plugins/dashboard/public/dashboard_top_nav/internal_dashboard_top_nav.tsx`?

Reply:
Ah ok - refactored in [841b2a8](https://github.com/elastic/kibana/pull/198619/commits/841b2a87d6918f81def0ff5c37e34ada91942d05)